### ISSUE TITLE
[mesheryctl] UX: Need more informative inactive token message

### ISSUE THREAD

**Issue by leecalcote** [meshery/meshery#4604]
_See [mesheryctl Command Tracker](https://bit.ly/3dqXy1q) for current status of commands._

#### Current Behavior
Users of `mesheryctl perf profile`, who don't have an active token receive the following message:

```
➜  ~ mesheryctl perf profile
Error: failed to make a request.Get "/": stopped after 10 redirects.
See https://docs.meshery.io/reference/mesheryctl/perf/profile for usage details
```

This behavior is true of a number of other commands. Each of the `app`, `pattern`, `mesh`, perf`.... should all apply the same logic and UX to the user - https://github.com/meshery/meshery/blob/2bf19d36cfd924c9c2e07fbb8686c975689309c5/mesheryctl/internal/cli/root/system/check.go#L489

#### Desired Behavior
Users should be informed of their inactive token and informed to run `mesheryctl system login`.

#### Acceptance Tests
- [ ] 1. All `perf` commands apply this same check and UX.
- [ ] 2. All `mesh` commands apply this same check and UX.
- [ ] 3. All `app` commands apply this same check and UX.
- [ ] 4. All `pattern` commands apply this same check and UX.

---

#### Contributor Resources

- ⌨️ [mesheryctl Contributing Guide](https://github.com/meshery/meshery/blob/master/mesheryctl/README.md)
- [Beginner's guide to contributing to Meshery and mesheryctl](https://youtu.be/hh_kFLZx3G4)
- [mesheryctl Command Tracker](https://docs.google.com/spreadsheets/d/1q63sIGAuCnIeDs8PeM-0BAkNj8BBgPUXhLbe1Y-318o/edit#gid=0)
- [Meshery CLI Commands and Documentation](https://docs.google.com/document/d/1xRlFpElRmybJ3WacgPKXgCSiQ2poJl3iCCV1dAalf0k/edit#heading=h.5fucij4hc5wt)
- 🙋🏾🙋🏼 Questions: [Layer5 Discussion Forum](https://discuss.layer5.io) and [Layer5 Community Slack](http://slack.layer5.io)

**Comment by cpepper96** [comment]
Hi @leecalcote can I take a crack at this?

**Comment by leecalcote** [comment]
@cpepper96, yes, that'd be great. 👍

**Comment by navendu-pottekkat** [comment]
Should we send back a response from the Meshery server if the token is invalid?

The [current logic](https://github.com/meshery/meshery/issues/4604) is written with Meshery UI as its primary client I guess.

If we can get a response from the server that the token is invalid, we can handle it appropriately in mesheryctl and provide a more contextual error message.

Other than that, I don't see how we can handle this error without manually checking for this "stopped after 10 redirects" error.

@utkarsh-pro @Revolyssup @leecalcote what do you guys think we should do here?

**Comment by piyushsingariya** [comment]
@navendu-pottekkat is right. Else we'll have to add a step to verify if the error contains the phrase `stopped after 10 redirects`

**Comment by Revolyssup** [comment]
@navendu-pottekkat  @piyushsingariya  When an invalid token is seen, a redirection to /provider takes place with status code of Redirection , i.e 302. And golang internally has a infinite for loop which breaks when redirects cross 10. Since we are not UI clients, it will keep redirecting.So maybe we can add a check on mesheryctl that if the status code is 302, then this redirection issue is there. Which means we can safely assume that "Token was invalid."
So after each request , I would advise to add an addtional check that if statusCode==302 {throw the invalid token error}

**Comment by piyushsingariya** [comment]
@Revolyssup This error is thrown from the step where we perform the request, in other words error thrown before checking the statusCode.

**Comment by Revolyssup** [comment]
> @Revolyssup This error is thrown from the step where we perform the request, Or I can say in other words, this error thrown before checking the statusCode.

Yeah I know tht, that's what I'm saying. Before checking the error, check if the statuscode is 302.and send the appropriate response

**Comment by Revolyssup** [comment]
> > @Revolyssup This error is thrown from the step where we perform the request, Or I can say in other words, this error thrown before checking the statusCode.
> 
> Yeah I know tht, that's what I'm saying. Before checking the error, check if the statuscode is 302.and send the appropriate error, i.e. invalid token

**Comment by navendu-pottekkat** [comment]
Checking the status code makes sense. @Revolyssup Is it safe to assume that it would only redirect for the case where the token is invalid?

**Comment by Revolyssup** [comment]
> Checking the status code makes sense. @Revolyssup Is it safe to assume that it would only redirect for the case where the token is invalid?

Currently in meshery server that redirection takes place at the auth handler layer. So with the current implementation. Yes the /provider redirect takes place there only when the token is invalid. It never reaches the internal middle wares or any of the handlers.

**Comment by navendu-pottekkat** [comment]
Okay. @Revolyssup So we could check that in the CLI I guess. But should we handle this more elegantly and mention that the token is expired from the server itself instead of relying on the status code for the redirect?

**Comment by Revolyssup** [comment]
@navendu-pottekkat  I also found some more bugs. One for eg is, when provider is neither Meshery nor Local. Then even though the token is invalid. It says pattern applied, and this because we actually get a 200 response along with the login page HTML.

So let me explain the flow in few cases-
1. Provider=Meshery ,token=Invalid. This time, the error is caught in authmiddleware and a redirect is made to /provider. There again, the requests crosses the provider middleware safely and is redirected to /user/login which is then handled by InitiateLogin handler method.This time , because this is a redirect, there is no set cookie for token , hence a redirect is made to "/", from where the cycle continues as the request keeps bouncing between /user/login and /. The http library in go limits the redirects upto 10, hence after 10 redirects the error is thrown. So on this level , yes I would say checking for 302 is enough.

2. When the provider passed is neither Meshery nor local. provider is found as nil and request is redirected to /provider which returns the provider UI html along with status code 200. The solution to avoid this can be this:
		_, err = html.Parse(res.Body)
		if err == nil {
			log.Info("pattern could not be applied. Server returned HTML")
			return nil
		}

I think we can wrap response handling from meshery server in a single utility function and then reuse it to have a common behavior across all commands for different status codes and responses.
// @leecalcote

**Comment by navendu-pottekkat** [comment]
There is [this function](https://github.com/navendu-pottekkat/meshery/blob/9bd1ed2d6669c3fb571a6e06a77338b867b50c25/mesheryctl/internal/cli/root/perf/profile.go#L163) in mesheryctl that checks if the response is HTML and we do actually check for it in the `perf` commands.

The utility function you mentioned would check the status code plus the functionality of the one I linked here right?

And it would be used by the clients to identify such errors?

**Comment by cpepper96** [comment]
Sorry been traveling for the holidays and then was sick a few days as well. Going to make this a priority for the new year! @Mr-DG-Wick

**Comment by leecalcote** [comment]
Thank you, @cpepper96

**Comment by cpepper96** [comment]
![profileOP](https://user-images.githubusercontent.com/80851353/147864801-f2ac5bb9-889f-4f1f-a5b3-e2dec1fee9be.PNG)
So this is the output I'm getting when I run the command 'mesheryctl perf profile'. I'm a little confused is this not already the desired output? @leecalcote @navendu-pottekkat

**Comment by navendu-pottekkat** [comment]
@cpepper96 We hit this issue when the token is present but it is expired/invalid. A similar error message should be added to that scenario as well.

**Comment by piyushsingariya** [comment]
> ![profileOP](https://user-images.githubusercontent.com/80851353/147864801-f2ac5bb9-889f-4f1f-a5b3-e2dec1fee9be.PNG) So this is the output I'm getting when I run the command 'mesheryctl perf profile'. I'm a little confused is this not already the desired output? @leecalcote @navendu-pottekkat

Yes, but there are a few times when we get this error instead of the desired one.
```
Error: failed to make a request.Get "/": stopped after 10 redirects.
```
